### PR TITLE
Add support for metal_connection mode (tunnel)

### LINKED ISSUE #129: metal_connection mode support
Connections with dedicated type will have support for a mode field. This issue is a placeholder for that work.

Depends on:
* https://github.com/packethost/packngo/pull/285
* https://github.com/packethost/packngo/pull/270
* https://github.com/equinix/terraform-provider-metal/pull/137

### LINKED ISSUE #134: Implement tags for connection and virtualcircuit resources
The [Equinix Metal API spec](https://api.equinix.com/metal/v1/api-docs/) includes `tags` for connection and virtualcircuit resources. These have not been included in this provider.

### LINKED ISSUE #144: crash when importing resource metal_virtual_circuit
```
❯ terraform import metal_virtual_circuit.connection_attachment uuid
metal_virtual_circuit.connection_attachment: Importing from ID "uuid"...
metal_virtual_circuit.connection_attachment: Import prepared!
  Prepared metal_virtual_circuit for import
metal_virtual_circuit.connection_attachment: Refreshing state... [id=uuid]
╷
│ Error: Plugin did not respond
│ 
│ The plugin encountered an error, and failed to respond to the plugin.(*GRPCProvider).ReadResource call. The plugin logs may contain more details.
╵

 


Stack trace from the terraform-provider-metal_v2.1.0 plugin:

 

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x18d03b8]

 

goroutine 81 [running]:
github.com/equinix/terraform-provider-metal/metal.resourceMetalVirtualCircuitRead(0xc0003284d0, 0x19d8080, 0xc000528500, 0xc0003284d0, 0x0)
        github.com/equinix/terraform-provider-metal/metal/resource_metal_virtual_circuit.go:134 +0x2d8
github.com/hashicorp/terraform-plugin-sdk/helper/schema.(*Resource).RefreshWithoutUpgrade(0xc00050fb00, 0xc000092b90, 0x19d8080, 0xc000528500, 0xc000326120, 0x0, 0x0)
        github.com/hashicorp/terraform-plugin-sdk@v1.0.0/helper/schema/resource.go:455 +0x12e
github.com/hashicorp/terraform-plugin-sdk/internal/helper/plugin.(*GRPCProviderServer).ReadResource(0xc000508100, 0x1c0e290, 0xc0003fcf90, 0xc0000929b0, 0xc000508100, 0xc0003fcf90, 0xc0005c3a50)
        github.com/hashicorp/terraform-plugin-sdk@v1.0.0/internal/helper/plugin/grpc_provider.go:525 +0x3dd
github.com/hashicorp/terraform-plugin-sdk/internal/tfplugin5._Provider_ReadResource_Handler(0x1a74460, 0xc000508100, 0x1c0e290, 0xc0003fcf90, 0xc000500840, 0x0, 0x1c0e290, 0xc0003fcf90, 0xc000314180, 0xb3)
        github.com/hashicorp/terraform-plugin-sdk@v1.0.0/internal/tfplugin5/tfplugin5.pb.go:3153 +0x214
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0000cc000, 0x1c16478, 0xc000602480, 0xc000310000, 0xc00009d620, 0x20e3e70, 0x0, 0x0, 0x0)
        google.golang.org/grpc@v1.23.0/server.go:995 +0x482
google.golang.org/grpc.(*Server).handleStream(0xc0000cc000, 0x1c16478, 0xc000602480, 0xc000310000, 0x0)
        google.golang.org/grpc@v1.23.0/server.go:1275 +0xd2c
google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc0003fa000, 0xc0000cc000, 0x1c16478, 0xc000602480, 0xc000310000)
        google.golang.org/grpc@v1.23.0/server.go:710 +0xab
created by google.golang.org/grpc.(*Server).serveStreams.func1
        google.golang.org/grpc@v1.23.0/server.go:708 +0xa5

Error: The terraform-provider-metal_v2.1.0 plugin crashed!
```

_Reported by @christianhxc_

### PR DESCRIPTION
**Pull request by displague**
* [x] Fixes #144 
* [x] Fixes #129 
* [x] Fixes #134
* [x] nni_vlan is optional for virtual circuits (when mode=tunnel, leave it to the API to enforce) (confirm this)

### CODE REVIEW

File: metal/resource_metal_connection.go [hunk 1]
```diff
@@ -53,6 +55,12 @@ func resourceMetalConnection() *schema.Resource {
 				Description: "Connection type - dedicated or shared",
 				ForceNew:    true,
 			},
+			"mode": {
```
Comment (t0mk) [inline metal/resource_metal_connection.go:58]: You could add a validator:

```
ValidateFunc: validation.StringInSlice([]string{"standard", "tunnel"}, false),
```

If you don't mind, also add it to `type` and `redundancy`

File: metal/resource_metal_virtual_circuit.go [hunk 2]
```diff
@@ -78,13 +78,17 @@ func resourceMetalVirtualCircuitCreate(d *schema.ResourceData, meta interface{})
...
 	}
+
 	connId := d.Get("connection_id").(string)
 	portId := d.Get("port_id").(string)
 	projectId := d.Get("project_id").(string)
 
+	if nniVlan, ok := d.GetOk("nni_vlan"); ok {
+		vncr.NniVLAN = nniVlan.(int)
+	}
+
```
Comment (t0mk) [inline metal/resource_metal_virtual_circuit.go:91]: Shall we do a sanity check for when user set nni_vlan but connection mode is not `tunnel`?

**Comment by t0mk** [comment]
@displague I fixed some stuff, in this pr, see f45071c

One question: Is it possible that `mode == "tunnel"` is valid only for dedicated connection, not for a shared one? The API seems to set the `tunnel` only for dedicated connection.

**Comment by displague** [comment]
> One question: Is it possible that mode == "tunnel" is valid only for dedicated connection, not for a shared one? The API seems to set the tunnel only for dedicated connection.

Correct. Only dedicated allows `tunnel` mode.

**Comment by displague** [comment]
The testing account did not access to the connection mode feature. I've enabled it and started the tests up again.

**Pull request merged**